lib/app/dispatcher.rb

```diff
@@ -133,6 +133,8 @@ def start_consumer!
           min_threads: MIN_THREADS,
           max_threads: MAX_THREADS,
           max_queue: MAX_QUEUE,
+          max_ingestion_queue_size: (App::Config.max_ingestion_queue_size || Utility::Constants::DEFAULT_MAX_INGESTION_QUEUE_SIZE).to_i,
+          max_ingestion_queue_bytes: (App::Config.max_ingestion_queue_bytes || Utility::Constants::DEFAULT_MAX_INGESTION_QUEUE_BYTES).to_i,
```

Comment:
I would suggest moving them to the top, together with other constants, e.g. `MAX_QUEUE`.

Reply:
A fun thing with it - it makes it impossible to mock it properly because the assignment happens when the file is included, thus I cannot have a "natural" test that sets `App::Config.max_ingestion_queue_size` first to nil, then to a value, do you know a way around it?

Reply:
Will `stub_const` work?

Reply:
`stub_const` will just mock the constant, so it will ignore whatever happens in `App::Config`. 

By having `App::Config.poll_interval` be nil or non-nil, you are essentially testing the ruby operator `||`.

Reply:
I'm trying to test a logical feature that says "if poll_interval is provided in config, then it's respected and used in scheduler properly, if it's not provided, then a different default value is used".

I don't particularly like it this way here, IMO we need to move the defaulting logic to configuration class and I can remove these tests from here

Reply:
Like this:
```ruby
context 'constant' do
  let(:poll_interval) { 120 }
  before(:each) do
    allow(App::Config).to receive(:poll_interval).and_return(poll_interval)
  end

  it 'uses configured poll interval' do
    expect(described_class.const_get('POLL_INTERVAL')).to eq(poll_interval)
  end
  
  context 'when it\`s not configured' do
    let(:poll_interval) { nil }

    it 'uses the default poll interval' do
      expect(described_class.const_get('POLL_INTERVAL')).to eq(3)
    end
  end
end
```

Again, I don't really think it's necessary.

>if poll_interval is provided in config, then it's respected and used in scheduler properly, if it's not provided, then a different default value is used

IMO, when you test `scheduler`, you shouldn't care whether the `poll_interval` is coming from configured value or default.

Reply:
I've tried similar approach to the one you described in the snippet, and the problem is `POLL_INTERVAL` is assigned once for the whole run-time of the tests in the batch, only when the class `App::Dispatcher` is loaded.

Therefore only one test of ones that you mentioned will work - depending on which one will be executed first OR if another file already loaded `App::Dispatcher` it will depend on a setup of the test that loaded it.

I don't see a way around it without redesigning the module and defaulting logic.

Reply:
> the problem is `POLL_INTERVAL` is assigned once for the whole run-time of the tests in the batch, only when the class `App::Dispatcher` is loaded.

OK, I never tested the snippet above, so I'm not sure if it works or not. My point is we really don't need to test `App::Config.poll_interval || 3`. Not a big deal, we can always refactor later.